src/Microsoft.Diagnostics.NETCore.Client/ReversedServer/ReversedDiagnosticsServer.cs

```diff
@@ -0,0 +1,201 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Diagnostics.NETCore.Client
+{
+    /// <summary>
+    /// Establishes server endpoint for processes to connect when configured to provide diagnostics connection is reverse mode.
+    /// </summary>
+    internal sealed class ReversedDiagnosticsServer : IDisposable
+    {
+        // The amount of time to allow parsing of the advertise data before cancelling. This allows the server to
+        // remain responsive in case the advertise data is incomplete and the stream is not closed.
+        private static readonly TimeSpan ParseAdvertiseTimeout = TimeSpan.FromSeconds(1);
+
+        private readonly CancellationTokenSource _cancellation = new CancellationTokenSource();
+        private readonly ConcurrentDictionary<Guid, ServerIpcEndpoint> _endpoints = new ConcurrentDictionary<Guid, ServerIpcEndpoint>();
```

Comment:
The initial comments for the PR aimed for a two layer design where the higher level ConnectionSource would manage a pool of clients, but some of the client management responsibility seems to be in ReversedDiagnosticsServer too. As best I can tell these endpoints are serving a few roles but there are alternatives that wouldn't force us to maintain and synchronize duplicated lists of most client information.
a) they cache connections that AcceptAsync() has already notified its caller about to avoid double notifications. Given that the higher layer already has a list of clients it is aware of this filtering seems unnecessary.
b) the endpoints store the cached stream and StreamTargets for pending operations. In the experimental refactor I moved that state to a Dictionary<Guid, Stream> and a List of pending operations.